monty_profiling/floppy/flop_counting/operations/arithmetic.py

```diff
@@ -0,0 +1,71 @@
+import numpy as np
+import warnings
+from typing import Any, Optional, Tuple
+
+__all__ = ["Addition", "Subtraction", "Multiplication", "Division"]
+
+class ArithmeticOperation:
+    """Base class for arithmetic operations."""
+
+    def __init__(self, name: str):
+        self.name = name
+
+    def count_flops(self, *args: Any, result: Any) -> int:
+        """Counts the floating point operations (FLOPs) for an arithmetic operation.
+
+        Handles both scalar and array operations. For scalar operations, counts one FLOP
+        per element in the non-scalar array. For array operations, counts one FLOP per
+        element in the result array, taking into account broadcasting.
+
+        Args:
+            *args: Variable length argument list containing the input operands.
+                Expected to have at least 2 arguments where each can be a scalar
+                or numpy array.
+            result: The result of the arithmetic operation, used to determine the
+                final shape after broadcasting.
+
+        Returns:
+            int: The total number of FLOPs performed in the operation, calculated as:
+                - For scalar operations: size of the non-scalar array
+                - For array operations: product of the result array's dimensions
+        """
+        # Handle scalar operations
+        if np.isscalar(args[0]) or np.isscalar(args[1]):
+            array_arg = args[1] if np.isscalar(args[0]) else args[0]
+            return np.size(array_arg)
+
+        # Handle array operations
+        arrays = [np.asarray(arg) for arg in args[:2]]
+        # Get the result shape directly from the actual result
+        # This ensures we correctly account for broadcasting
+        result_shape = np.asarray(result).shape
+
+        return np.prod(result_shape)
+
+
+class Addition(ArithmeticOperation):
+    """Class for addition operation."""
+
+    def __init__(self):
+        super().__init__("add")
+
+
+class Subtraction(ArithmeticOperation):
+    """Class for subtraction operation."""
+
+    def __init__(self):
+        super().__init__("subtract")
+
+
+class Multiplication(ArithmeticOperation):
+    """Class for multiplication operation."""
+
+    def __init__(self):
+        super().__init__("multiply")
+
+
+class Division(ArithmeticOperation):
```

Comment:
I noticed while doing some further reading on FLOPs that division is sometimes considered >1 (e.g. 4) FLOPs, due to differences in how it is handled in hardware. However, it is also often estimated as 1. I think it would be worth adding a reference somewhere to the following as justification for our use of 1. 

https://www.stat.cmu.edu/~ryantibs/convexopt-F18/scribes/Lecture_19.pdf